universal-hash/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+//! Traits for [Universal Hash Functions].
+//!
+//! Universal hash functions select from a "universal family" of possible
+//! hash functions selected by a key. They are well suited to the purpose
+//! of "one time authenticators" for a sequence of bytestring inputs,
+//! as their construction has a number of desirable properties such as
+//! pairwise independence as well as amenability to efficient implementations,
+//! particularly when implemented using SIMD instructions.
+//!
+//! When combined with a cipher, such as in Galois/Counter Mode or the
+//! Salsa20 family AEAD constructions, they can provide the core functionality
+//! for a Message Authentication Code (MAC).
+
+#![no_std]
+#![doc(html_logo_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo_small.png")]
+#![warn(missing_docs)]
+
+#[cfg(feature = "std")]
+extern crate std;
+
+pub extern crate generic_array;
+extern crate subtle;
+
+use generic_array::{ArrayLength, GenericArray};
+use subtle::{Choice, ConstantTimeEq};
+
+/// The `UniversalHash` trait defines a generic interface for universal hash
+/// functions.
+pub trait UniversalHash<BlockSize: ArrayLength<u8>>: Clone
+where
+    BlockSize::ArrayType: Copy,
+{
+    /// Instantiate a universal hash function with the given key
+    fn new(key: Block<BlockSize>) -> Self;
+
+    /// Input a block into the universal hash function
+    fn input_block(&mut self, block: Block<BlockSize>);
+
+    /// Input data into the universal hash function. If the length of the
+    /// data is not a multiple of the block size, the remaining data is
+    /// padded with zeroes up to the `BlockSize`.
+    ///
+    /// This approach is frequently used by AEAD modes which use
+    /// Message Authentication Codes (MACs) based on universal hashing.
+    fn input_padded(&mut self, data: &[u8]) {
+        let block_size = BlockSize::to_usize();
+
+        for chunk in data.chunks(block_size) {
```

Comment:
You can write a bit more efficient code like this:
```rust
let mut chunks = data.chunks_exact(block_size);
for chunk in &mut chunks {
    let block_bytes = GenericArray::clone_from_slice(chunk);
    self.input_block(block_bytes.into());
}
let rem = chunks.remainder();
if !rem.is_empty() {
    let mut padded_block = GenericArray::default();
    padded_block[..chunk.len()].copy_from_slice(rem);
    self.input_block(padded_block.into());
}
```

Reply:
Most of that looks good, however...

```
GenericArray::clone_from_slice(chunk)
```

...you're calling `Clone` on a (XMM register-sized) `Copy`  type

Reply:
IIUC `clone()` for `Copy` types is as efficient as a simple `memcpy`, so performance-wise `GenericArray::clone_from_slice(chunk)` and `*GenericArray::from_slice(chunk)` should be equivalent.

Reply:
Plus here you still working with bytes, conversion is done later with `into()`.

Reply:
I guess I can get rid of the `Copy` bound for now, but when I didn't have it on both Poly1305 and POLYVAL it resulted in more `movups` instructions whereas using a `Copy` type let them accumulate directly into an XMM register.

Reply:
(note: everything I'm describing was with regular `[u8; 16]`s. Not sure what effect `GenericArray` would have on that)

Reply:
We also could use `GenericArray::from_slice(chunk)` and change `fn input_block(&mut self, block: Block<BlockSize>);` to `fn input_block(&mut self, block: &GenericArray<u8, BlockSize>);`. BTW maybe it's worth to name it as something like `block_update` in the light of the future renaming of methods?

Reply:
Changing my other impls from `&[u8; 16]` to `[u8; 16]` resulted in ~5% speedup, which is what I was trying to preserve here.

I used `input_*` for consistency, but if you'd like to just go strait to `update_*` that's fine by me.

Reply:
Hm, this is strange. From `for` loop you essentially get a single pointer and it shouldn't matter where it will be read. Can you repeat benchmarks but with trait this time? Don't forget to mark `input_block` with `#[inline]`.

>I used input_* for consistency, but if you'd like to just go straight to update_* that's fine by me.

I think it's fine to per-emptively rename it now, it will be less stuff to update later.

Reply:
I think I'd just rather drop the `Copy` bound, for simplicity and consistency's sake (so we can switch back to an associated type)